### PR TITLE
Most basic agent modelling APIs in a stepped simulation + discrete event stepping

### LINKED ISSUE #36: Discrete Event Simulation for simulation container

### LINKED ISSUE #37: Docs Simulation

### PR DESCRIPTION
**Pull request by rcschrg**
* World for the SimulationContainer
* Space as part of the world
* stepping hooks
* discrete event stepping
closes #36 
closes #37

### CODE REVIEW

File: test/simulation_container_tests.jl [hunk 1]
```diff
@@ -293,4 +294,81 @@ end
     @test agent1.scheduled_counter == 121
     @test agent2.counter == 2
     @test agent2.scheduled_counter == 200
+end
+
+@testset "SimulationWithSpecificDelaysAndScheduledTasksWithReplyOnHandleDiscreteEvent" begin
```
Comment (jsagerOffis) [inline test/simulation_container_tests.jl:299]: Does this name really have to be this long?

File: test/simulation_container_tests.jl [hunk 2]
```diff
@@ -293,4 +294,81 @@ end
...
+    #=
+    using Logging
+    debug_logger = ConsoleLogger(stderr, Logging.Debug)
+    global_logger(debug_logger)
+    =#
```
Comment (jsagerOffis) [inline test/simulation_container_tests.jl:304]: remove commented out block?

File: src/container/simulation.jl [hunk 3]
```diff
@@ -56,7 +74,7 @@ function to_message_package(message_tuple::Tuple{Any, AbstractDict}, simulation_
     return MessagePackage(sender_aid, receiver_aid, simulation_time, message_tuple)
 end
 
-function to_cs_input(message_queue::ConcurrentQueue{Tuple{Any,AbstractDict,DateTime}})::Vector{MessagePackage}
+function to_cs_input!(message_queue::ConcurrentQueue{Tuple{Any,AbstractDict,DateTime}})::Vector{MessagePackage}
```
Comment (jsagerOffis) [inline src/container/simulation.jl:77]: `ConcurrentQueue{Tuple{Any,AbstractDict,DateTime}}`

Do we gain anything by moving `Tuple{Any,AbstractDict,DateTime}` into its own struct for readability here and in other places this is used?

File: src/simulation/tasks.jl [hunk 4]
```diff
@@ -25,6 +25,9 @@ end
 function step_iteration(task_sim::TaskSimulation, step_size_s::Real)::TaskIterationResult 
     throw(ErrorException("Please implement step_iteration(...)"))
 end
+function determine_next_event_time(task_sim::TaskSimulation)
```
Comment (jsagerOffis) [inline src/simulation/tasks.jl:28]: should probably be covered by a test

File: src/simulation/tasks.jl [hunk 5]
```diff
@@ -40,6 +43,37 @@ struct WaitResult
...
 
+function determine_next_event_time(scheduler::SimulationScheduler, simulation_time::DateTime)
+    lowest = nothing
+
+    # normal queue
+    next = scheduler.queue.head.next
+    while !isnothing(next)
+        if isa(next.value, Tuple)
+            return 0
+        else
```
Comment (jsagerOffis) [inline src/simulation/tasks.jl:54]: should definitely be covered by a test

**Comment by rcschrg** [comment]
@jsagerOffis done, please check

**Review by jsagerOffis — APPROVED** [review]

**Pull request merged**